components/api_version/src/api_v2.rs

```diff
@@ -107,4 +107,58 @@ impl APIVersion for APIV2 {
         value.user_value.push(flags.bits());
         value.user_value
     }
+
+    fn decode_raw_key(encoded_key: &Key, with_ts: bool) -> Result<(Vec<u8>, Option<TimeStamp>)> {
+        debug_assert!(verify_encoded_key(encoded_key, with_ts));
+        let ts = decode_raw_key_timestamp(encoded_key, with_ts)?;
+        Ok((encoded_key.to_raw()?, ts))
+    }
+
+    fn decode_raw_key_owned(
+        encoded_key: Key,
+        with_ts: bool,
+    ) -> Result<(Vec<u8>, Option<TimeStamp>)> {
+        debug_assert!(verify_encoded_key(&encoded_key, with_ts));
+        let ts = decode_raw_key_timestamp(&encoded_key, with_ts)?;
+        Ok((encoded_key.into_raw()?, ts))
+    }
+
+    fn encode_raw_key(user_key: &[u8], ts: Option<TimeStamp>) -> Key {
+        debug_assert_eq!(Self::parse_key_mode(user_key), KeyMode::Raw);
+        let encoded_key = Key::from_raw(user_key);
+        if let Some(ts) = ts {
+            encoded_key.append_ts(ts)
+        } else {
+            encoded_key
+        }
+    }
+
+    // `encode_raw_key_owned` has improvement only for APIV1|APIV1TTL, but not for APIV2.
+    // As there is no "in place" implementation available for encoding.
+    fn encode_raw_key_owned(user_key: Vec<u8>, ts: Option<TimeStamp>) -> Key {
+        Self::encode_raw_key(&user_key, ts)
+    }
+}
+
+#[inline]
+fn verify_encoded_key(encoded_key: &Key, with_ts: bool) -> bool {
+    let encoded_bytes = &encoded_key.as_encoded()[..];
+    if !matches!(APIV2::parse_key_mode(encoded_bytes), KeyMode::Raw) {
+        return false;
+    }
+    let mut len = encoded_bytes.len();
+    if with_ts {
+        len = len.checked_sub(number::U64_SIZE).unwrap();
+    }
+    len > 0 && (len % (tikv_util::codec::bytes::ENC_GROUP_SIZE + 1) == 0)
```

Comment:
How about `bytes::decode_bytes(&mut encoded_bytes, false).is_ok() && encoded_bytes.len() == 8 * (with_ts as usize)`? So it doesn't need to depend on the details of codec.